File: backend/app/schemas/system/tenant.py
"""
企业相关 Schema / Tenant Schema

定义企业 API 的请求和响应数据结构
Defines tenant API request and response data structures.
"""

from datetime import datetime
from typing import Any

from pydantic import Field, model_validator

from app.core.base_schema import BaseSchema
from app.schemas.tenant.domain import TenantDomainSimpleResponse

_RETIRED_TENANT_INPUT_FIELDS = frozenset({"plan", "settings"})


def _reject_retired_tenant_input_fields(data: Any) -> Any:
    """中文: 企业写入只接受新契约字段，退役字段进入请求即拒绝。

    EN: Tenant writes accept only new-contract fields; retired fields are
    rejected on input.
    """
    if isinstance(data, dict):
        retired = _RETIRED_TENANT_INPUT_FIELDS & set(data.keys())
        if retired:
            fields = ", ".join(sorted(retired))
            raise ValueError(f"{fields} is retired; use plan_id or ConfigService")
    return data


class TenantPlanInfo(BaseSchema):
    """企业套餐信息（简略） / Tenant plan info (brief)."""

    id: int = Field(..., description="套餐 ID")
    code: str = Field(..., description="套餐代码")
    name: str = Field(..., description="套餐名称")


class TenantStorageStats(BaseSchema):
    """企业存储统计信息 / Tenant storage statistics."""

    used_bytes: int = Field(0, description="已使用存储空间（字节）")
    limit_bytes: int = Field(0, description="存储限制（字节），0 表示无限制")
    limit_gb: int = Field(0, description="存储限制（GB），0 表示无限制")
    usage_percent: float = Field(0.0, description="使用率百分比")
    file_count: int = Field(0, description="附件总数")
    unlimited: bool = Field(True, description="是否无限制")


class TenantResponse(BaseSchema):
    """企业信息响应 / Tenant info response."""

    id: int = Field(..., description="企业 ID")
    code: str = Field(..., description="企业编码")
    name: str = Field(..., description="企业名称")
    contact_name: str | None = Field(None, description="联系人姓名")
    contact_phone: str | None = Field(None, description="联系人电话")
    contact_email: str | None = Field(None, description="联系人邮箱")
    is_active: bool = Field(..., description="是否启用")
    # 套餐信息（新版） / Plan info (new)
    plan_id: int | None = Field(None, description="套餐 ID")
    plan_info: TenantPlanInfo | None = Field(None, description="套餐信息")
    quota: dict[str, Any] | None = Field(None, description="配额配置")
    expires_at: datetime | None = Field(None, description="到期时间")
    remark: str | None = Field(None, description="备注")
    # 域名信息 / Domains
    primary_domain: TenantDomainSimpleResponse | None = Field(
        None, description="主域名"
    )
    domains: list[TenantDomainSimpleResponse] = Field(
        default_factory=list, description="域名列表"
    )
    domain_count: int = Field(0, description="域名数量")
    # 存储配额信息 / Storage quota
    storage_stats: TenantStorageStats | None = Field(None, description="存储统计信息")
    # 时间字段 / Timestamps
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    @model_validator(mode="before")
    @classmethod
    def extract_tenant_info(cls, data: Any) -> Any:
        """从 ORM 关系中提取套餐和域名信息 / Extract plan and domain info from ORM relations."""
        # 如果是 ORM 对象，转换为字典 / If ORM object, convert to dict
        if hasattr(data, "__table__"):
            result = {
                "id": data.id,
                "code": data.code,
                "name": data.name,
                "contact_name": data.contact_name,
                "contact_phone": data.contact_phone,
                "contact_email": data.contact_email,
                "is_active": data.is_active,
                "plan_id": data.plan_id,
                "quota": data.quota,
                "expires_at": data.expires_at,
                "remark": data.remark,
                "created_at": data.created_at,
                "updated_at": data.updated_at,
            }

            # 提取套餐信息 / Extract plan
            if hasattr(data, "tenant_plan") and data.tenant_plan is not None:
                plan = data.tenant_plan
                result["plan_info"] = {
                    "id": plan.id,
                    "code": plan.code,
                    "name": plan.name,
                }

            # 提取域名信息 / Extract domains
            if hasattr(data, "domains") and data.domains:
                domains_list = [
                    {
                        "id": d.id,
                        "domain": d.domain,
                        "domain_type": d.domain_type,
                        "is_primary": d.is_primary,
                        "is_verified": d.is_verified,
                        "ssl_status": d.ssl_status,
                    }
                    for d in data.domains
                    if not d.is_deleted
                ]
                result["domains"] = domains_list
                result["domain_count"] = len(domains_list)
                # 找到主域名 / Pick primary domain
                result["primary_domain"] = next(
                    (d for d in domains_list if d["is_primary"]), None
                )

            return result
        return data


class TenantCreateRequest(BaseSchema):
    """创建企业请求 / Create tenant request."""

    name: str = Field(..., min_length=1, max_length=100, description="企业名称")
    contact_name: str | None = Field(None, max_length=50, description="联系人姓名")
    contact_phone: str | None = Field(None, max_length=20, description="联系人电话")
    contact_email: str | None = Field(None, description="联系人邮箱")
    # 套餐 ID（新版）
    plan_id: int | None = Field(None, description="套餐 ID")
    quota: dict[str, Any] | None = Field(
        None, description="配额配置（可覆盖套餐默认值）"
    )
    expires_at: datetime | None = Field(None, description="到期时间")
    remark: str | None = Field(None, max_length=500, description="备注")
    # 企业超级管理员账号（必填） / Tenant super admin (required)
    admin_username: str = Field(
        ..., min_length=2, max_length=50, description="管理员用户名"
    )
    admin_email: str = Field(..., description="管理员邮箱")
    admin_password: str = Field(
        ..., min_length=6, max_length=100, description="管理员密码"
    )

    @model_validator(mode="before")
    @classmethod
    def reject_retired_fields(cls, data: Any) -> Any:
        return _reject_retired_tenant_input_fields(data)


class TenantUpdateRequest(BaseSchema):
    """更新企业请求 / Update tenant request."""

    name: str | None = Field(None, min_length=1, max_length=100, description="企业名称")
    contact_name: str | None = Field(None, max_length=50, description="联系人姓名")
    contact_phone: str | None = Field(None, max_length=20, description="联系人电话")
    contact_email: str | None = Field(None, description="联系人邮箱")
    # 套餐 ID（新版）
    plan_id: int | None = Field(None, description="套餐 ID")
    quota: dict[str, Any] | None = Field(
        None, description="配额配置（可覆盖套餐默认值）"
    )
    expires_at: datetime | None = Field(None, description="到期时间")
    remark: str | None = Field(None, max_length=500, description="备注")

    @model_validator(mode="before")
    @classmethod
    def reject_retired_fields(cls, data: Any) -> Any:
        return _reject_retired_tenant_input_fields(data)


class TenantStatusRequest(BaseSchema):
    """企业状态切换请求 / Tenant status toggle request."""

    is_active: bool = Field(..., description="是否启用")


class TenantImpersonateRequest(BaseSchema):
    """一键登录企业后台请求 / Impersonate tenant admin request."""

    role_id: int | None = Field(None, description="目标角色 ID（可选）")


class TenantImpersonateResponse(BaseSchema):
    """一键登录企业后台响应 / Impersonate tenant admin response."""

    impersonate_token: str = Field(
        ..., description="一键登录 Token（60秒有效，一次性）"
    )
    tenant_code: str = Field(..., description="企业编码")
    tenant_name: str = Field(..., description="企业名称")
    expires_in: int = Field(60, description="Token 有效期（秒）")


class TenantResetOwnerPasswordRequest(BaseSchema):
    """重置企业超级管理员密码请求 / Reset tenant super admin password request."""

    new_password: str = Field(..., min_length=6, max_length=100, description="新密码")


__all__ = [
    "TenantPlanInfo",
    "TenantStorageStats",
    "TenantResponse",
    "TenantCreateRequest",
    "TenantUpdateRequest",
    "TenantStatusRequest",
    "TenantImpersonateRequest",
    "TenantImpersonateResponse",
    "TenantResetOwnerPasswordRequest",
]

File: frontend/apps/web-antd/src/api/admin/tenant.ts
/**
 * 租户管理 API
 * 对接后端 /admin/tenants/* 接口
 */
import type { ApiRequestOptions } from '#/utils/request';

import { requestClient } from '#/utils/request';

// ============================================================
// 类型定义
// ============================================================

/** 套餐类型 */
export type TenantPlan = 'basic' | 'enterprise' | 'free' | 'pro';

/** 租户列表查询参数 */
export type TenantListParams = Record<string, unknown>;

/** 创建租户请求 */
export interface TenantCreateRequest {
  /** 租户编码（可选，后端自动生成） */
  code?: string;
  name: string;
  contact_name?: null | string;
  contact_phone?: null | string;
  contact_email?: null | string;
  plan?: TenantPlan;
  quota?: null | Record<string, any>;
  expires_at?: null | string;
  remark?: null | string;
}

/** 更新租户请求 */
export interface TenantUpdateRequest {
  name?: null | string;
  contact_name?: null | string;
  contact_phone?: null | string;
  contact_email?: null | string;
  plan?: null | TenantPlan;
  quota?: null | Record<string, any>;
  expires_at?: null | string;
  remark?: null | string;
}

/** 切换状态请求 */
export interface TenantStatusRequest {
  is_active: boolean;
}

/** 域名简要信息（后端原始格式） */
export interface TenantDomainBriefRaw {
  id: number;
  tenant_id: number;
  domain: string;
  is_verified: boolean;
  verified_at: null | string;
  is_primary: boolean;
  ssl_status: 'active' | 'failed' | 'pending';
  ssl_expires_at: null | string;
  cname_target: null | string;
  remark: null | string;
  created_at: string;
  updated_at: string;
}

/** 域名简要信息（前端格式） */
export interface TenantDomainBrief {
  id: number;
  domain: string;
  domainType: 'custom' | 'default';
  isPrimary: boolean;
  verificationStatus: 'pending' | 'verified';
  sslStatus: 'active' | 'failed' | 'pending';
  cnameTarget?: null | string;
}

/** 套餐简要信息（后端原始格式） */
export interface TenantPlanBriefRaw {
  id: number;
  code: string;
  name: string;
}

/** 套餐简要信息（前端格式） */
export interface TenantPlanBrief {
  id: number;
  code: string;
  name: string;
}

/** 租户信息（后端原始格式 snake_case） */
export interface TenantInfoRaw {
  id: number;
  code: string;
  name: string;
  contact_name?: string;
  contact_phone?: string;
  contact_email?: string;
  plan?: null | TenantPlan;
  plan_id?: null | number;
  plan_info?: null | TenantPlanBriefRaw;
  quota?: Record<string, any>;
  is_active: boolean;
  expires_at?: string;
  remark?: string;
  created_at: string;
  updated_at?: string;
  // 域名信息
  primary_domain?: null | TenantDomainBriefRaw;
  domain_count?: number;
  domains?: TenantDomainBriefRaw[];
}

/** 租户信息（前端格式 camelCase） */
export interface TenantInfo {
  id: number;
  code: string;
  name: string;
  contactName?: string;
  contactPhone?: string;
  contactEmail?: string;
  plan?: null | TenantPlan;
  planId?: null | number;
  planInfo?: null | TenantPlanBrief;
  quota?: Record<string, any>;
  isActive: boolean;
  expiresAt?: string;
  remark?: string;
  createdAt: string;
  updatedAt?: string;
  // 域名信息
  primaryDomain?: null | TenantDomainBrief;
  domainCount?: number;
  domains?: TenantDomainBrief[];
}

/** 分页列表响应 */
export interface TenantListResponse {
  items: TenantInfo[];
  total: number;
  page: number;
  page_size: number;
}

// ============================================================
// 转换函数
// ============================================================

/** 转换域名简要信息 */
function transformDomainBrief(raw: TenantDomainBriefRaw): TenantDomainBrief {
  return {
    id: raw.id,
    domain: raw.domain,
    // 根据 remark 或其他逻辑推断类型，因为后端未返回 domain_type
    domainType: raw.remark?.includes('默认') ? 'default' : 'custom',
    isPrimary: raw.is_primary,
    verificationStatus: raw.is_verified ? 'verified' : 'pending',
    sslStatus: raw.ssl_status,
    cnameTarget: raw.cname_target,
  };
}

/** 将后端 snake_case 转换为前端 camelCase */
function transformTenantInfo(raw: TenantInfoRaw): TenantInfo {
  return {
    id: raw.id,
    code: raw.code,
    name: raw.name,
    contactName: raw.contact_name,
    contactPhone: raw.contact_phone,
    contactEmail: raw.contact_email,
    plan: raw.plan,
    planId: raw.plan_id,
    planInfo: raw.plan_info || null,
    quota: raw.quota,
    isActive: raw.is_active,
    expiresAt: raw.expires_at,
    remark: raw.remark,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
    // 域名信息
    primaryDomain: raw.primary_domain
      ? transformDomainBrief(raw.primary_domain)
      : null,
    domainCount: raw.domain_count,
    domains: raw.domains?.map((d) => transformDomainBrief(d)),
  };
}

// ============================================================
// API 接口
// ============================================================

const API_PREFIX = '/admin/tenants';

/** 租户下拉选项结果 */
export interface TenantSelectOption {
  label: string;
  value: number;
  extra?: {
    code?: string;
    isActive?: boolean;
  };
}

/**
 * 获取租户下拉选项
 * GET /admin/tenants/select
 *
 * 权限: tenant:select
 */
export async function getTenantSelectApi(
  params?: Record<string, unknown>,
  options?: ApiRequestOptions,
): Promise<{ items: TenantSelectOption[] }> {
  return requestClient.get<{ items: TenantSelectOption[] }>(
    `${API_PREFIX}/select`,
    { params, ...options },
  );
}

/**
 * 获取租户列表
 * GET /admin/tenants
 */
export async function getTenantListApi(
  params?: TenantListParams,
  options?: ApiRequestOptions,
): Promise<TenantListResponse> {
  const response = await requestClient.get<{
    items: TenantInfoRaw[];
    page: number;
    page_size: number;
    total: number;
  }>(API_PREFIX, { params, ...options });

  return {
    items: response.items.map((item) => transformTenantInfo(item)),
    total: response.total,
    page: response.page,
    page_size: response.page_size,
  };
}

/**
 * 获取租户详情
 * GET /admin/tenants/{tenant_id}
 */
export async function getTenantDetailApi(
  tenantId: number,
  options?: ApiRequestOptions,
): Promise<TenantInfo> {
  const raw = await requestClient.get<TenantInfoRaw>(
    `${API_PREFIX}/${tenantId}`,
    options,
  );
  return transformTenantInfo(raw);
}

/**
 * 创建租户
 * POST /admin/tenants
 */
export async function createTenantApi(
  data: TenantCreateRequest,
  options?: ApiRequestOptions,
): Promise<TenantInfo> {
  const raw = await requestClient.post<TenantInfoRaw>(
    API_PREFIX,
    data,
    options,
  );
  return transformTenantInfo(raw);
}

/**
 * 更新租户
 * PUT /admin/tenants/{tenant_id}
 */
export async function updateTenantApi(
  tenantId: number,
  data: TenantUpdateRequest,
  options?: ApiRequestOptions,
): Promise<TenantInfo> {
  const raw = await requestClient.put<TenantInfoRaw>(
    `${API_PREFIX}/${tenantId}`,
    data,
    options,
  );
  return transformTenantInfo(raw);
}

/**
 * 删除租户
 * DELETE /admin/tenants/{tenant_id}
 */
export async function deleteTenantApi(
  tenantId: number,
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.delete(`${API_PREFIX}/${tenantId}`, options);
}

/**
 * 切换租户状态
 * PUT /admin/tenants/{tenant_id}/status
 */
export async function toggleTenantStatusApi(
  tenantId: number,
  data: TenantStatusRequest,
  options?: ApiRequestOptions,
): Promise<TenantInfo> {
  const raw = await requestClient.put<TenantInfoRaw>(
    `${API_PREFIX}/${tenantId}/status`,
    data,
    options,
  );
  return transformTenantInfo(raw);
}

/** 重置租户管理员密码请求 */
export interface ResetTenantOwnerPasswordRequest {
  new_password: string;
}

/**
 * 重置租户管理员密码
 * PUT /admin/tenants/{tenant_id}/reset-owner-password
 */
export async function resetTenantOwnerPasswordApi(
  tenantId: number,
  data: ResetTenantOwnerPasswordRequest,
  options?: ApiRequestOptions,
): Promise<void> {
  await requestClient.put(
    `${API_PREFIX}/${tenantId}/reset-owner-password`,
    data,
    options,
  );
}

// ============================================================
// 一键登录相关
// ============================================================

/** 一键登录请求 */
export interface TenantImpersonateRequest {
  role_id?: null | number;
}

/** 一键登录响应 */
export interface TenantImpersonateResponse {
  impersonateToken: string;
  tenantCode: string;
  tenantName: string;
  expiresIn: number;
}

/** 后端原始响应 */
interface TenantImpersonateResponseRaw {
  impersonate_token: string;
  tenant_code: string;
  tenant_name: string;
  expires_in: number;
}

/**
 * 一键登录租户后台
 * POST /admin/tenants/{tenant_id}/impersonate
 * 生成一键登录 Token（60秒过期，一次性使用）
 */
export async function tenantImpersonateApi(
  tenantId: number,
  data?: TenantImpersonateRequest,
  options?: ApiRequestOptions,
): Promise<TenantImpersonateResponse> {
  const raw = await requestClient.post<TenantImpersonateResponseRaw>(
    `${API_PREFIX}/${tenantId}/impersonate`,
    data || {},
    options,
  );
  return {
    impersonateToken: raw.impersonate_token,
    tenantCode: raw.tenant_code,
    tenantName: raw.tenant_name,
    expiresIn: raw.expires_in,
  };
}
